### PR TITLE
[SPARK-35567][SQL] Fix: Explain cost is not showing statistics for all the nodes

### PR DESCRIPTION
**Pull request by shahidki31**
### What changes were proposed in this pull request?
Explain cost command in spark currently doesn't show statistics for all the nodes. It misses some nodes in almost all the TPCDS queries.
In this PR, we are collecting all the plan nodes including the subqueries and computing  the statistics for each node, if it doesn't exists in stats cache,

### Why are the changes needed?
**Before Fix**
For eg: Query1,  Project node doesn't have statistics
image

Query15, Aggregate node doesn't have statistics

image

**After Fix:** 
Query1:
image
Query 15:
image
### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual testing

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @HyukjinKwon @maropu @cloud-fan @srowen Kindly review

**Comment by srowen** [comment]
I don't know enough to review this, sorry

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -256,13 +255,9 @@ class QueryExecution(
...
     // trigger to compute stats for logical plans
     try {
-      optimizedPlan.foreach(_.expressions.foreach(_.foreach {
-        case subqueryExpression: SubqueryExpression =>
-          // trigger subquery's child plan stats propagation
-          subqueryExpression.plan.stats
-        case _ =>
-      }))
-      optimizedPlan.stats
+      optimizedPlan.collectWithSubqueries {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:258]: cc @cloud-fan @maryannxue FYI

**Comment by cloud-fan** [comment]
Does this fix nested subqueries?

**Comment by shahidki31** [comment]
@cloud-fan Yes, `collectWithSubqueries` will include nested subqueries as well.
image

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 2]
```diff
@@ -256,13 +255,9 @@ class QueryExecution(
 
     // trigger to compute stats for logical plans
     try {
-      optimizedPlan.foreach(_.expressions.foreach(_.foreach {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: how about 
```
// trigger subquery's child plan stats propagation
optimizedPlan.subqueriesAll.foreah(_.stats)
optimizedPlan.stats
```

Comment (shahidki31) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: @cloud-fan I have made changes like above and I think this doesn't fix the issue. Issue here is not all the nodes have stats in statsCache. 
This is from query3, which doesn't have subqueries, The aggregate node doesn't show stats with the above changes, but fix with the PR change

```
GlobalLimit 100, Statistics(sizeInBytes=2.9 KiB, rowCount=68)
+- LocalLimit 100, Statistics(sizeInBytes=3.5 KiB, rowCount=68)
   +- Sort [d_year#9 ASC NULLS FIRST, sum_agg#89 DESC NULLS LAST, brand_id#87 ASC NULLS FIRST], true, Statistics(sizeInBytes=3.5 KiB, rowCount=68)
      +- Aggregate [d_year#9, i_brand#62, i_brand_id#61], [d_year#9, i_brand_id#61 AS brand_id#87, i_brand#62 AS brand#88, MakeDecimal(sum(UnscaledValue(ss_ext_sales_price#45)),17,2) AS sum_agg#89]
         +- Project [d_year#9, ss_ext_sales_price#45, i_brand_id#61, i_brand#62], Statistics(sizeInBytes=14.0 MiB, rowCount=2.77E+5)
            +- Join Inner, (d_date_sk#3 = ss_sold_date_sk#53), Statistics(sizeInBytes=16.1 MiB, rowCount=2.77E+5)
               :- Project [ss_ext_sales_price#45, ss_sold_date_sk#53, i_brand_id#61, i_brand#62], Statistics(sizeInBytes=14.0 MiB, rowCount=2.77E+5)
......
```

Comment (shahidki31) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: Basically performing `optimizedPlan.stats` doesn't update `statsCache` for all the nodes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 3]
```diff
@@ -256,13 +255,9 @@ class QueryExecution(
...
-        case subqueryExpression: SubqueryExpression =>
-          // trigger subquery's child plan stats propagation
-          subqueryExpression.plan.stats
-        case _ =>
-      }))
-      optimizedPlan.stats
+      optimizedPlan.collectWithSubqueries {
+        case plan => plan.stats
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: I see, so this basically triggers stats calculation for every node in the plan tree.

Comment (shahidki31) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: Yes, right.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:259]: Can we add a comment to explain why we need to do it? It's not that obvious to many reviewers.

Comment (shahidki31) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:262]: Yes, I added a comment.

**Review by cloud-fan — APPROVED** [review]

**Comment by shahidki31** [comment]
Retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Comment by shahidki31** [comment]
Thanks a lot @cloud-fan

**Pull request closed without merging**